features/dd-sdk-android-flags/src/main/kotlin/com/datadog/android/flags/featureflags/internal/DatadogFlagsClient.kt

```diff
@@ -73,144 +75,356 @@ internal class DatadogFlagsClient(
+     * @return The JSON object value of the flag, or the default value if the flag cannot be resolved for any reason.
      */
     override fun resolveStructureValue(flagKey: String, defaultValue: JSONObject): JSONObject =
-        getValue(flagKey, defaultValue) { stringValue ->
-            try {
-                JSONObject(stringValue)
-            } catch (e: JSONException) {
-                featureSdkCore.internalLogger.log(
-                    level = InternalLogger.Level.ERROR,
-                    target = InternalLogger.Target.USER,
-                    messageBuilder = { ERROR_PARSING_JSON.format(Locale.US, flagKey) },
-                    throwable = e
+        resolveValue(flagKey, defaultValue)
+
+    /**
+     * Resolves a flag value by chaining internal resolution and tracking.
+     *
+     * This helper method simplifies the convenience methods by encapsulating the common pattern
+     * of resolving a flag internally and then processing it through the tracking layer.
+     *
+     * @param T The type of the flag value
+     * @param flagKey The name of the flag to query
+     * @param defaultValue The default value to return if resolution fails
+     * @return The resolved value or the default value
+     */
+    private fun <T> resolveValue(flagKey: String, defaultValue: T): T =
+        resolveTracked(resolveInternal(flagKey, defaultValue))
+
+    private fun writeExposureEvent(name: String, data: PrecomputedFlag, context: EvaluationContext) {
+        processor.processEvent(
+            flagName = name,
+            context = context,
+            data = data
+        )
+    }
+
+    private fun logEvaluation(key: String, value: Any) {
+        rumEvaluationLogger.logEvaluation(
+            flagKey = key,
+            value = value
+        )
+    }
+
+    /**
+     * Resolves a flag value with detailed resolution information.
+     *
+     * This is the core resolution method that centralizes all flag resolution logic:
+     * 1. Retrieves the precomputed flag from the repository
+     * 2. Returns FLAG_NOT_FOUND error if the flag doesn't exist
+     * 3. Parses the flag value based on the target type (T)
+     * 4. Returns TYPE_MISMATCH/PARSE_ERROR if parsing fails
+     * 5. Returns success with variant, reason, and metadata if parsing succeeds
+     *
+     * Type-specific parsing:
+     * - Boolean: Strict parsing (only "true"/"false")
+     * - String: No conversion needed
+     * - Int: Parse to integer
+     * - Double: Parse to double
+     * - JSONObject: Parse JSON with error logging
+     *
+     * @param T The type of the flag value (Boolean, String, Int, Double, or JSONObject).
+     * @param flagKey The name of the flag to query.
+     * @param defaultValue The value to return if the flag cannot be retrieved or parsed.
+     * @return [ResolutionDetails] with either the parsed value and metadata, or an error.
+     */
+    override fun <T> resolve(flagKey: String, defaultValue: T): ResolutionDetails<T> {
```

Comment:
minor: it is better to move this up, so that we have interface methods grouped first and only after them we have `private` methods.

Reply:
done